docs/2_2_1_fingerprinting.md

```diff
@@ -65,116 +65,244 @@ It is used to uniquely match a method by its characteristics.
 Fingerprinting is used to match methods with a limited amount of known information.
 Methods with obfuscated names that change with each update are primary candidates for fingerprinting.
 The goal of fingerprinting is to uniquely identify a method by capturing various attributes, such as the return type,
-access flags, an opcode pattern, strings, and more.
+access flags, instructions, strings, and more.
 
 ## ⛳️ Example fingerprint
 
-An example fingerprint is shown below:
-
 ```kt
-
-package app.revanced.patches.ads.fingerprints
-
-fingerprint {
+val showAdsFingerprint by fingerprint {
+    // Method signature.
     accessFlags(AccessFlags.PUBLIC, AccessFlags.FINAL)
     returns("Z")
-    parameters("Z")
-    opcodes(Opcode.RETURN)
-    strings("pro")
-    custom { (method, classDef) -> classDef == "Lcom/some/app/ads/AdsLoader;" }
+    // Declared parameters are matched using String.startsWith()
+    // Non obfuscated classes should be declared using the full class name.
+    // While obfuscated class names must be declared only using the object type
+    // Since obfuscated names change between releases.
+    // Last parameter is simply `L` since it's an obfuscated class object.
+    parameters("Ljava/lang/String;", "I", "L")
+    
+    // Method implementation:
+    instructions( 
+        // Filter 1.
+        fieldAccess(
+            definingClass = "this",
+            type = "Ljava/util/Map;"
+        ),
+
+        // Filter 2.
+        string("showBannerAds"),
+      
+        // Filter 3.
+        methodCall(
+            definingClass = "Ljava/lang/String;",
+            name = "equals",
+        ),
+
+        // Filter 4.
+        // maxAfter = 0 means this must match immediately after the last filter.
+        opcode(Opcode.MOVE_RESULT, maxAfter = 0),
```

Comment:
This is odd api. If opcode has a maxAfter, why does literal(), or string() or fieldAccess() not have maxAfter? Consistency is important! MaxAfter implies MaxBefore, MinAfter and MinBefore as well. Either all or none

Reply:
All instruction filters have `maxAfter`.

The maxAfter is very much required for opcodes, so the old `opcodes {}` declaration works.

I'm open to naming this differently. I tried a few different names for this parameter but settled on this because the previous names were too wordy or confusing.

I also tried other ways to represent variable space between filters, such as an explicite "variable space" filter, but all attempts were too complicated or too confusing.